Add CMake build system (#220)
libusb/hidapi: CMakeLists.txt

> cmake_minimum_required(VERSION 3.1.3 FATAL_ERROR)

Author: > nobody is on cmake <3 anymore

You will be surprised.

Although, define "legacy". It is not deprecated, and it is not recommended by CMake developers not to use.

Effectively, without it we only fail to give one good error message, in case if someone _accidentally_ tries to use a very old CMake.

> cmake_minimum_required(VERSION 3.1.3 FATAL_ERROR)

Reviewer: None of the other functions in the CMake file would work if they were on <3. That's why you don't see it in modern CMake configs anymore.

> cmake_minimum_required(VERSION 3.1.3 FATAL_ERROR)

Author: > None of the other functions in the CMake file would work if they were on <3

That is incorrect. Strictly speaking: "_some_ of the functions in this build tree will not work, _or will have a different outcome, that the script author intended_"

project(hidapi LANGUAGES C)

if(NOT DEFINED CMAKE_BUILD_TYPE OR CMAKE_BUILD_TYPE STREQUAL "")
    set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Debug;Release;MinSizeRel;RelWithDebInfo" FORCE)
endif()

> set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")

Reviewer 2: What is the point of this? Why don't you leave it to CMake to set the default?

Author: > What is the point of this? 

It is for convenience of those, who are building hidapi as a standalone package/library. Try cmake-gui and you'll see.

> Why don't you leave it to CMake to set the default?

Because by default CMake doesn't set it at all. Maybe it does makes sence somtime, but I haven't seen a single real project where it would.

---
As for the convenience part: see my other comment https://github.com/libusb/hidapi/pull/220#discussion_r541672007

Reviewer 2: I still don't get the point of overriding CMake's default behavior, but I guess it doesn't hurt? :shrug: It'll just be a bit of confusing code for people reading this in the future.

Author: > but I guess it doesn't hurt?

That's my point.

> CMake's default behavior

Humor me. Can you tell what's CMake's default behavior if you don't specify CMAKE_BUILD_TYPE at all?

Reviewer 3: @Youw 
There are multi-config generators (like MSVC) that ignores `CMAKE_BUILD_TYPE`. The value of this variable is also not limited to the ones you've listed - for example I use also `TSan` and `ASan` build type for sanitizers. I agree this part of code isn't necessary here but I don't think it will break something on the other hand.

Author: > ignores `CMAKE_BUILD_TYPE`

The problem that it is not exactly ignored. It is still available in the CMake scripts, and what bothers me the most: CMake doesn't default it to empty value when MSVC generator is used, but rather to "Debug" (gcc + make - it is defaulted to empy value).
I find such behavior way more confusing, but I guess CMake's developers must have had their reasons for that.

> I use also `TSan` and `ASan` build type for sanitizers

I must admit - I haven't seen any custom CMAKE_BUILD_TYPE usage/implementation, so I guess that is a good point too.
Can you point to me a reference how your/typical usage look like?

As I mentioned earlier, a standalone build likely not to be used with a non-standard build types but I agree that prohibiting it wouldn't be a good thing, so there is [that](https://github.com/libusb/hidapi/commit/8ae8ffb2cf0ef6596baa55a9c8db3fc2dfc31e9f).

> 

Author: so instead of type less, you suggest always type more (and possible type more errors)
I vote for user-friendly

> muscle memory for me personally

and it makes it harder to use for those, who don't have such muscle memory

---
even more: I use cmake-gui even more often then a command-line. It saves me much time. And it saves me time to pick a selection from a combo-box, instead of typing it manually each time.

> 

Reviewer: > so instead of type less, you suggest always type more

In this case, yes. It's not about typing less/more. It's about not colliding with CMake builtins in the future.

Further, most people that will be using this will be either installing from a package manager or using it as a submodule - and only the latter would require someone typing it once to configure it.

> and it makes it harder to use for those, who don't have such muscle memory

The point wasn't that I'm better. The point is that it's a pretty standard way to use CMake.

> even more: I use cmake-gui even more often then a command-line

That is not the majority case.

> 

Author: > or using it as a submodule - and only the latter would require someone typing it once to configure it.

And what problem does it make? Does it limit the users?

> The point wasn't that I'm better. 

Neither was mine.

> The point is that it's a pretty standard way to use CMake.

For all my "small" project I have a different muscle memory: `cmake .. && make -j`. Pretty standard to me (and I didn't invented it).

> 

Reviewer: The point here is that this is not common. Another user now has asked about it; you generally don't set these things, because debug/release isn't binary. I've rarely, hardly ever actually, seen a config that does this. It's the general case yo specify the config type you want to build for.

if(APPLE)

>     option(CMAKE_FRAMEWORK "Build macOS/iOS Framework version of the library" OFF)

Author: you're mistaken, this is not a HIDAPI option, this is a [standard CMake option](https://cmake.org/cmake/help/latest/variable/CMAKE_FRAMEWORK.html), and it is selected into an `option` for convenience of those, who like cmake-gui as much as I do

>     option(CMAKE_FRAMEWORK "Build macOS/iOS Framework version of the library" OFF)

Reviewer: You're right, google didn't bring it up for some reason. Ignore.

elseif(NOT WIN32)
    if(CMAKE_SYSTEM_NAME MATCHES "Linux")
        option(HIDAPI_WITH_HIDRAW "Build HIDRAW-based implementation of HIDAPI" ON)

>         option(HIDAPI_WITH_LIBUSB "Build LIBUSB-based implementation of HIDAPI" ON)

Reviewer 2: Is the libusb backend usable on Windows?

Author: Theoretically speaking - it _could be_ usable.
Practically: its implementation relies on POSIX build environment.

I believe it is possible to replace all unix/linux-specific dependencies and replace it with C11, but that is not done yet.

    endif()

> endif()

Author: > Linux and no other *NIX distributions

I believe this is exactly what is originally written

> option(BUILD_SHARED_LIBS "When set: build shared version of the libraries, otherwise - static" ON)

Author: you're absolutely right
again, without it - cmake-gui doesn't give a nice hint about it

> option(BUILD_SHARED_LIBS "When set: build shared version of the libraries, otherwise - static" ON)

Reviewer: Fair enough.

> option(BUILD_SHARED_LIBS "When set: build shared version of the libraries, otherwise - static" ON)

Reviewer 2: ```suggestion
option(BUILD_SHARED_LIBS "Build shared version of the libraries, otherwise build statically" ON)
```

> set(HIDAPI_INSTALL_TARGETS ON)

Author: unfortunately you don't know the design that is hidden behind it

in short - the root CMakeLists.txt is not designed to be a subdirectory of a larger project `src/CMakeLists.txt` is designed for that
I'll surely explain that in the documentation/usage examples as soon as I can get my hands to it

> set(HIDAPI_INSTALL_TARGETS ON)

Reviewer: >  the root CMakeLists.txt is not designed to be a subdirectory of a larger project

It should be. That's how CMake libraries should be structured. The fact they can be installed system-wide is an extra thing, not the default.

You cut out users that want to statically link without polluting their system cache, or those who are cross compiling, etc.

> set(HIDAPI_INSTALL_TARGETS ON)

Author: > It should be. That's how CMake libraries should be structured.

I don't suppose that statement is based on large number of different real-life CMake-based projects/libraries.

---
I had a pleasure integrating maybe a dozen CMake-based libraries into a larger CMake-based solution/build tree (as a submodule). And none of those projects could be integrated by simple `add_subdirectory(<subproject root>)`, and I don't believe that is because those projects or their build systems are badly implemented, not at all.
Although, a solution/build tree I mentioned above has lots of demands, some of them: it has to compile/deploy on at least 4 platforms: Windows, macOS, Linux, Android, including host x86_64 builds and ARM64 cross-compilations. All these and other requirements automatically applies to all dependencies.

Per my observations, each projects's build system consist of of two logical parts: _build infrastructure_ and _build definitions_. (It is not always possible to tell exactly where the infrastructure ends, and the definitions begins, but still.)
CMake itself is an infrastructure. It is designed to suit as much different possible needs as possible. That also means that it will never work of-the-box for everyone. CMake has many [variables](https://cmake.org/cmake/help/latest/manual/cmake-variables.7.html) and [properties](https://cmake.org/cmake/help/latest/manual/cmake-properties.7.html) to configure the build system, but it doesn't mean, that any combination of those values/properties can by applied to any project successfully. So each project (its author(s)) choose a subset of its variables/properties, maybe its values, into an actual CMake options that are intended to be configurable/changed. That all together makes a build infrastructure, that answers to question "how?": how to build a target, which compiler options, where to install, what kind of the binary should be, etc.
At this point we didn't even start developing our project, we don't have build definitions yet, we don't have our "what". Q: _what_ do we want to? - Make a library. So we add a build definition (a target): `add_library(...)`. There are certain properties of each target, that are absolutely required to successfully build a target, e.g.: macOS backend of hidapi has to be linked against `IOKit framework`, otherwise of won't build at all. That is essential, that has to be part of build definition. It cannot be optional or configurable.
On the other hand, we can build our source with `-O2` or `-g -O0` compiler options, but that doesn't change the essence/intention of the project. Maybe performance, at most, maybe convenience for the developer (e.g. ability to debug), and that is a part of the project's infrastructure.

My point is that root CMakeLists.txt usually contains a lot of infrastructure for convenience of the project's developer. Because project's developer keeps its focus on the ability to integrate the project into a package manager, install into a system (and make the library available for other projects), deploy it to production, ability to develop it fast and efficient.
hidapi is not an exception here. All the options are designed for _maximum convenience of the user/developer_, who tries to build hidapi as a _standalone project/library_. Install project or not - that is a question to the user: call/build `install` target or not, it shouldn't be a question to the build system as "add `install` target or not?".
I tried to cover all possible cases I could thing of. Most CMake-based projects I've seen so far, are actually doing the same in one degree or another. Without a good infrastructure, a usage of the project's build system is not impossible, but less convenient/harder for sure.

On the other hand, when someone wan'ts to use a hidapi as a subdirectory of the existing CMake build tree, it likely means they _already have its own build infrastructure_. CMake options that _make perfect sense for hidapi_ might have absolutely _no sense to the other project_ that want's to use hidapi (e.g. as exactly a single static library). So the only thing other project needs from hidapi - its _build definitions_, and minimum dependencies.
By current design, root CMakeLists.txt of hidapi contains an _infrastructure_. `src/CMakeLists.txt` and below - that's definitions. As an author of the build system, I've tried to design it that way, and only because I've thought of all that even before I've started the development.
Unfortunately not all of the infrastructure can be removed from definitions (at least, because some of the infrastructure is actually what makes definitions possible), so all of the infrastructure parts that are left are designed to be configurable as much as possible, to cover the majority of possible usages.

> set(HIDAPI_INSTALL_TARGETS ON)

Reviewer 4: You can detect whether you're the main project or included as a subproject as follows:
```cmake
cmake_minimum_required(VERSION 3.0)
if(PROJECT_SOURCE_DIR)
    message(STATUS "Included as a subproject")
    set(HIDAPI_MAIN_PROJECT OFF)
else()
    message(STATUS "I'm the main project")
    set(HIDAPI_MAIN_PROJECT ON)
endif()
project(libusb)
```

That way, there's no need for two `CMakeLists.txt` files.

> set(HIDAPI_INSTALL_TARGETS ON)

Author: The two `CMakeLists.txt` files are still there, just because the defaults/behavior will be a bit different, when built as a sub_directory and as a standalone project, but yes, this idea is already implemented: https://github.com/libusb/hidapi/commit/81291b205b52013019c2498620b6d27f2ee3bcc9

And technically, checking for `PROJECT_SOURCE_DIR` isn't the same as checking for `CMAKE_/CURRENT_/SOURCE_DIR`, and isn't strictly correct. No one prohibits you to do something like:
```cmake
cmake_minimum_required(VERSION 3.0)

add_subdirectory(hidapi)

project(my_awsome_project)

#...
```

add_subdirectory(src)

set(BUILD_HIDTEST_DEFAULT OFF)
if(CMAKE_BUILD_TYPE STREQUAL "Debug")
    set(BUILD_HIDTEST_DEFAULT ON)
endif()
option(BUILD_HIDTEST "Build small console test application hidtest" ${BUILD_HIDTEST_DEFAULT})
if(BUILD_HIDTEST)

>     add_subdirectory(hidtest)

Reviewer: ```suggestion
    enable_testing()
    add_subdirectory(hidtest)
```

Author: `hidtest` is not testing module, as I explained in my other comments

Reviewer: Ah okay.

cmake_minimum_required(VERSION 3.1.3 FATAL_ERROR)

if(NOT CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
    add_subdirectory(src)
    # compatinility with find_package() vs add_subdirectory
    set(hidapi_VERSION "${hidapi_VERSION}" PARENT_SCOPE)
    return()
endif()
# All of the below in this file is meant for a standalone build.

> # When building as a subdirectory of a larger project, most of the options may not make sence for it,

Reviewer 2: ```suggestion
# When building as a subdirectory of a larger project, most of the options may not make sense for it,
```
